[DOCS] Docs examples style checker

ChrisChinchilla opened the pull request:
### Description

Part of https://github.com/ethereum/solidity/issues/4581 and https://github.com/ethereum/solidity/issues/4580 this creates the test and config needed. Once we're happy with this, I have the files it highlighted and fixed ready too.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

ChrisChinchilla commented:
I had pylint installed and it did some formatting fixes, can undo those if not welcome

ChrisChinchilla commented on test/.solhintignore:
```diff
@@ -0,0 +1 @@
+*_style_guide_rst.sol
```
The Style guide file is challenging to check right now for a variety of reasons, so will just ignore for now.

chriseth commented:
Sorry, I think I just commented on the follow-up PR.

ChrisChinchilla commented on scripts/isolate_tests.py:
```diff
@@ -59,6 +59,7 @@ def extract_docs_cases(path):
 def write_cases(f, tests):
     cleaned_filename = f.replace(".","_").replace("-","_").replace(" ","_").lower()
     for test in tests:
+        remainder = re.sub(r'^ {4}', '', test, 0, re.MULTILINE)
```
@chriseth In reference to comments in other PR, I stripped this back to just the line I added and removed all auto linting.

leonardoalt replied:
What exactly is this doing?

ChrisChinchilla replied:
I wrote this some time ago 😅.

Err. All the code examples when extracted are indented by 8 spaces, which violates the style guide, so before checking you need to remove 4 from each line.

chriseth replied:
Please add that as a comment to the file.

ChrisChinchilla commented:
@chriseth Further tidying with respect to comments from https://github.com/ethereum/solidity/pull/6427

solhint has a quiet mode, but it's not very good. It still prints a blank line, but I'll log an issue with them and for no, hopefully this is OK for us.

chriseth commented:
Please squash.

chriseth commented:
```
Compiling all examples from the documentation...
test_003f37a8c0b062dada91446e685bf29e58e4d1c16b5c51f9f995a62874a9a912_style_guide_rst.sol
scripts/../test/cmdlineTests.sh: line 279: solhint: command not found
```

ChrisChinchilla commented:
Hmm, obviously this is because a dependency is missing, but this raises a question for me.

As the dependency is a node module, which also requires Node installed. A CI system likely does and we can just install the module. But a user _might_ not, though probably does.

I'm thinking that this test should be extracted a test that people can run by setting or unsetting a parameter (like many of the other tests), and then mention in the docs, that running the test requires node, and if it is selected, then install the NPM module.

Does that make sense? Thoughts?

chriseth commented:
Please squash.

ChrisChinchilla commented:
Done @chriseth but what about the comment above? We haven't really addressed the dependency issue yet.

chriseth commented:
I would suggest the following:
 - install solhint on the CI that runs the test
 - only execute the solhint line if solhint is available, otherwise print a warning but do not fail the test.

ChrisChinchilla commented:
@chriseth OK, that's a good idea, but now it opens another question for me, and helps me understand the decisions behind our test strategy a bit more.

I created this test in the CLI tests, not for CI tests, so it might make more sense to break the test out into a seperate file so it can be reused for CLI tests and as a CI test? And then implement your suggestions above?

chriseth commented on test/cmdlineTests.sh:
```diff
@@ -306,6 +305,22 @@ SOLTMPDIR=$(mktemp -d)
+            cp "$REPO_ROOT"/test/.solhint.json "$SOLTMPDIR"/.solhint.json
+        else
+            echo "node is installed, installing solhint"
+            npm install -g solhint
```
I'm a bit uneasy with installing software globally in a test script. Perhaps it would be better to just install it in the circle setup routine and just don't run it if it is not available?

ChrisChinchilla replied:
I am too. Right now I am mostly still trying to figure out the best way to implement this, and I can't experiment without pushing to Ci, which is slow… The biggest stumbling block so far has actually been installing node itself, but getting there.

chriseth commented on test/cmdlineTests.sh:
```diff
@@ -306,6 +305,21 @@ SOLTMPDIR=$(mktemp -d)
     cd "$SOLTMPDIR"
     "$REPO_ROOT"/scripts/isolate_tests.py "$REPO_ROOT"/docs/ docs
+
+    if node -v >/dev/null
```
```suggestion
    if node -v >/dev/null 2>&1
```

chriseth replied:
Also, why do you check for node if what you actually want is npm?

ChrisChinchilla replied:
That's a good point, I was still testing this tbh as waiting for each PR to run the tests each time is taking a while, some of these changes were not really final, but this is a good suggestion 👍

chriseth commented on test/cmdlineTests.sh:
```diff
@@ -315,6 +329,13 @@ SOLTMPDIR=$(mktemp -d)
         fi
         echo "$f"
+
+        if npm list -g | grep solhint >/dev/null
```
`npm` might not be installed.

chriseth commented on test/cmdlineTests.sh:
```diff
@@ -329,7 +350,7 @@ SOLTMPDIR=$(mktemp -d)
     done
 )
-rm -rf "$SOLTMPDIR"
+# rm -rf "$SOLTMPDIR"
```
```suggestion
rm -rf "$SOLTMPDIR"
```

chriseth commented on test/docsExamples.sh:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
File name: Perhaps add "style" to the name, or do you intend this file to test more at a later point in time?

leonardoalt replied:
Yea I thought the same, this way it sounds like it's actually testing the examples for compilation and not only style.

ChrisChinchilla replied:
Missed this comment, done

chriseth commented on test/docsExamples.sh:
```diff
@@ -0,0 +1,49 @@
+    set -e
+    cd "$SOLTMPDIR"
+
+    if node -v >/dev/null
```
This could be extracted into a bash utility file

ChrisChinchilla commented:
OK, @chriseth I think this is finally ready. I wanted to seperate things as this test is less important than others and didn't really want it reliant on on other more important tests, and needed time to break it out.

The tests in this PR will fail as the style fixes happened in https://github.com/ethereum/solidity/pull/6427 but either way, this test failing shouldn't break a build, more advisory.

Will tweak over time, and also add to docs.

ChrisChinchilla commented:
Rebased, but there are new violations to the style guide introduced after this PR that I am fixing.

leonardoalt commented:
Tests are failing

ChrisChinchilla commented:
Yeah, this has sat here a while @leonardoalt I'll rebase, but the only test failing is the one this introduces which should be fixed by a rebase.

ChrisChinchilla commented:
OK added an ignore step to ignore the contributors guide, as it features syntax that breaks the checker.

chriseth commented on test/docsExamples.sh:
```diff
@@ -0,0 +1,49 @@
+            echo "node is installed, but not solhint"
+        fi
+    else
+        echo "node not installed, skipping docs style checker"
```
Don't you need to break here? How can the `solhint` command below work if nodejs is not installed?

ChrisChinchilla commented:
@chriseth Added comment and refactored bash code so it breaks when either dependency not installed

chriseth commented on test/docsCodeStyle.sh:
```diff
@@ -0,0 +1,49 @@
+            done
+        else
+            echo "node is installed, but not solhint"
+            exit
```
```suggestion
            exit 1
```

chriseth commented on test/docsCodeStyle.sh:
```diff
@@ -0,0 +1,49 @@
+        fi)
+    else
+        echo "node not installed, skipping docs style checker"
+        exit
```
```suggestion
        exit 1
```

chriseth commented on test/docsCodeStyle.sh:
```diff
@@ -0,0 +1,49 @@
+    cd "$SOLTMPDIR"
+    "$REPO_ROOT"/scripts/isolate_tests.py "$REPO_ROOT"/docs/ docs
+
+    (if npm -v >/dev/null 2>&1; then
```
No need for `(`

Commit:
Add style checker

Rename files

Changes from review

Update test/docsCodeStyle.sh

Co-Authored-By: chriseth <chris@ethereum.org>
Update test/docsCodeStyle.sh

Co-Authored-By: chriseth <chris@ethereum.org>
Remove extraneous brackets

ChrisChinchilla commented:
@chriseth Squashed, and addressed comments, sorry some things I read said the brackets were needed when using break and exit, obviously that was wrong.
